Define `parse_version` for broad compatibility with `packaging`

goerz opened the pull request:
Closes #27

goerz committed:
Encapsulate `parse_version` function

This will allow us to write a `parse_version` function that is
compatible with versions of `packaging` both before and after v22.0

goerz committed:
Use `packvers` backport of `packaging`

This restores the `LegacyVersion` behavior, which we depend on.

ZLLentz commented:
I'll drop these links for the context on `packvers`:
https://github.com/pypa/packaging/issues/530#issuecomment-1362592132
https://github.com/pypa/packaging/issues/530#issuecomment-1362717017

I think it's a good solution since it covers all the current use cases and doesn't require a big rewrite. I'll be nice to be able to install this alongside packages that require the most recent `packaging` versions :+1:.

klauer commented:
I admittedly don't fully understand the need to retain the legacy version parsing. However as it seems to be a crucial part of your workflow, @goerz, I think this is a mutually-agreeable solution and am on board 👍 

By the way - we'll have to update the README indicating just how many repos use this library one day. In our organization alone, we have about 100 repositories using docs-versions-menu. Overall we're very happy with it. Thanks again for your great work here!

goerz committed:
Remove unused sort_key

goerz committed:
Introduce NonVersionFolderName for more clarity

goerz commented:
> I admittedly don't fully understand the need to retain the legacy version parsing. However as it seems to be a crucial part of your workflow, @goerz, I think this is a mutually-agreeable solution and am on board 👍

It's required because I'm putting all folder names in the `gh-pages` branch (or wherever the `docs-versions-menu`command runs) through that parser. This is to [sort and categorize](https://goerz.github.io/docs_versions_menu/v0.5.1/folderspecs.html#group-names) the folder names. The folder names will usually be the names of tags (released versions like `v0.1.0`) or branch names. The `LegacyVersion` specifically handles all branch names (`master`/`main`, or whatever other branches might be set up to generate documentation). Thus, the name `LegacyVersion` is actually a bit confusing: this isn't about some "legacy" versioning convention, but about arbitrary branch names that aren't versions at all. It's probably a good idea to make the intent a bit more explicit, and hence I've added https://github.com/goerz/docs_versions_menu/pull/32/commits/34489f62e5a119afe3e13c81bbcf1af81b026d2c.


> By the way - we'll have to update the README indicating just how many repos use this library one day. In our organization alone, we have about 100 repositories using docs-versions-menu. Overall we're very happy with it. Thanks again for your great work here!

Wow! I'm glad it's getting some use!

awoimbee commented on src/docs_versions_menu/parse_version.py:
```diff
@@ -0,0 +1,26 @@
+    pass
+
+
+def parse_version(version):
+    """Parse `version` string into either a `Version` or a
+    `NonVersionFolderName` object."""
+    try:
+        return packvers.version.Version(version)
+    except packvers.version.InvalidVersion:
+        return NonVersionFolderName(version)
```
Why not use `packvers.version.parse` (it would handle LegacyVersion for you; [ref](https://github.com/nexB/packvers/blob/main2/docs/version.rst#reference)) ?
Otherwise with this `try`/`except` you could continue using `packaging.version.parse` instead of `packvers`.

goerz replied:
I was trying to replace `LegacyVersion` with the `NonVersionFolderName` subclass of `LegacyVersion` that I'm defining to clarify that it's not really a "legacy version". You're right that I could use `packaging` or `packvers` interchangeably here, but I still need `packvers.version.LegacyVersion` to subclass from. The `packaging` package completely removed `LegacyVersion`, right? So, I still need the `packvers` dependency, and at that point I might as well use `packvers` for `parse`/`Version` instead of `packaging`.

awoimbee reviewed (commented):
Thanks !

goerz committed:
Inherit `NonVersionFolderName` from `_BaseVersion`

This allows us to go back to using `packaging` instead of `packvers` as
the dependency for parsing folder names/version numbers.

We'll have to depend on the internal `packaing.version._BaseVersion`,
but we can test the behavior that we rely on in case `packaging` makes a
breaking change to that behavior.

goerz committed:
Run CI against both `packaging < 22.0` and `>=22.0`

goerz commented:
Just for completeness, https://github.com/pypa/packaging/pull/669 is quite interesting and points to [`packaging_legacy`](https://github.com/di/packaging_legacy) as yet another workaround for the removal of `LegacyVersion`. Alas, I'm perfectly happy to side-step the whole drama and implement the required functionality in `docs_versions_menu` directly. Hopefully, `packaging` will be more stable in the future, but if not, I'll just keep moving stuff into here.

klauer reviewed (approved):
I understand the issue better now and like the approach you chose here, sticking with `packaging` and extending it for the purposes of `docs-versions-menu`.

If/when the upstream library API changes, it can be revisited then.

ZLLentz commented:
The final approach here is clean and the code's intent is clear, I like it a lot
